syntax = "proto3";
package ricedb;

service RiceDB {
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc CreateUser(CreateUserRequest) returns (CreateUserResponse);
  rpc DeleteUser(DeleteUserRequest) returns (DeleteUserResponse);
  rpc Health(HealthRequest) returns (HealthResponse);
  
  rpc Insert(InsertRequest) returns (InsertResponse);
  rpc GetNode(GetNodeRequest) returns (GetNodeResponse);
  rpc DeleteNode(DeleteNodeRequest) returns (DeleteNodeResponse);
  rpc Search(SearchRequest) returns (SearchResponse);
  rpc BatchInsert(stream InsertRequest) returns (BatchInsertResponse);

  rpc CreateSession(CreateSessionRequest) returns (CreateSessionResponse);
  rpc SnapshotSession(SnapshotSessionRequest) returns (SnapshotSessionResponse);
  rpc LoadSession(LoadSessionRequest) returns (LoadSessionResponse);
  rpc CommitSession(CommitSessionRequest) returns (CommitSessionResponse);
  rpc DropSession(DropSessionRequest) returns (DropSessionResponse);
}

message LoginRequest {
  string username = 1;
  string password = 2;
}

message LoginResponse {
  string token = 1;
  int64 userId = 2;
  string role = 3;
}

message CreateUserRequest {
  string username = 1;
  string password = 2;
  string role = 3;
}

message CreateUserResponse {
  int64 userId = 1;
  string message = 2;
}

message DeleteUserRequest {
  string username = 1;
}

message DeleteUserResponse {
  bool success = 1;
  string message = 2;
}

message HealthRequest {
}

message HealthResponse {
  string status = 1;
  string version = 2;
}

message InsertRequest {
  int64 id = 1;
  string text = 2;
  bytes metadata = 3;
  int64 userId = 4;
  string sessionId = 5;
  repeated float embedding = 6;
}

message InsertResponse {
  bool success = 1;
  int64 nodeId = 2;
  string message = 3;
}

message GetNodeRequest {
  int64 nodeId = 1;
  string sessionId = 2;
}

message GetNodeResponse {
  Node node = 1;
}

message Node {
  int64 id = 1;
  bytes metadata = 2;
}

message DeleteNodeRequest {
  int64 nodeId = 1;
  string sessionId = 2;
}

message DeleteNodeResponse {
  bool success = 1;
  string message = 2;
}

message SearchRequest {
  string queryText = 1;
  int64 userId = 2;
  int32 k = 3;
  string sessionId = 4;
  string filter = 5; 
  repeated float queryEmbedding = 6;
}

message SearchResponse {
  repeated SearchResult results = 1;
}

message SearchResult {
  int64 id = 1;
  float similarity = 2;
  bytes metadata = 3;
}

message BatchInsertResponse {
  int32 count = 1;
  repeated int64 nodeIds = 2;
}

message CreateSessionRequest {
  string parentSessionId = 1;
}

message CreateSessionResponse {
  string sessionId = 1;
}

message SnapshotSessionRequest {
  string sessionId = 1;
  string path = 2;
}

message SnapshotSessionResponse {
  bool success = 1;
}

message LoadSessionRequest {
  string path = 1;
}

message LoadSessionResponse {
  string sessionId = 1;
}

message CommitSessionRequest {
  string sessionId = 1;
  string mergeStrategy = 2;
}

message CommitSessionResponse {
  bool success = 1;
}

message DropSessionRequest {
  string sessionId = 1;
}

message DropSessionResponse {
  bool success = 1;
}
